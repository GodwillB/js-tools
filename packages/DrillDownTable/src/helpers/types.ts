import { Dictionary } from '@onaio/utils';
import { TableJSXProps } from '..';

export type DrillDownFilter<D extends object> = (
  props: DrillDownTableProps<D>,
  parentId: string
) => D[];

/** describes props for the DrillDownTable component */
export interface DrillDownTableProps<D extends object>
  extends Omit<TableJSXProps<D>, 'fetchData' | 'parentNodes'> {
  extraCellProps?: Dictionary /** props to be given to CellComponent */;
  CellComponent: React.ElementType /** The component used to render the cell that has the drill down */;
  loading: boolean /** if loading */;
  loadingComponent: React.ElementType /** custom component to show whilst loading is true */;
  drillDownFilter: DrillDownFilter<D>;
}
