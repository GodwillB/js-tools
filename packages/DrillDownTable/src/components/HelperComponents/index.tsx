import React from 'react';
import { Cell } from 'react-table';
import {
  CARET_SPAN,
  CLICKABLE_CSS_CLASS,
  LINKER_ITEM_CSS_CLASS,
  LOADING,
  NO_DATA_FOUND
} from '../../helpers/constants';

export interface NullDataComponentProps {
  noDataFound: string;
}

const defaultNullDataProps: NullDataComponentProps = {
  noDataFound: NO_DATA_FOUND
};

/** Rendered component when there is no data for table component to render */
const NullDataComponent = (props: NullDataComponentProps) => (
  <div className="jumbotron">
    <p>{props.noDataFound}</p>
  </div>
);

NullDataComponent.defaultProps = defaultNullDataProps;
export { NullDataComponent };

/** Interface for DropDown cell props */
export interface DropDownCellProps {
  cell: Cell;
  cellValue: Node;
  hasChildren: boolean;
}

/** Component that will be rendered in drop-down table cells showing a caret
 * that indicates if you can drill-down a row of data or not.
 */
export const DropDownCell: React.ElementType = (props: DropDownCellProps) => {
  const { cellValue, hasChildren } = props;
  return (
    <div className={hasChildren ? CLICKABLE_CSS_CLASS : LINKER_ITEM_CSS_CLASS}>
      <span>
        {cellValue}
        {hasChildren && CARET_SPAN}
      </span>
    </div>
  );
};

export interface SpinnerProps {
  loadingText: string;
}

const defaultSPinnerProps: SpinnerProps = {
  loadingText: LOADING
};

/** default loader component */
const Spinner = (props: SpinnerProps) => (
  <div className="spinner-border m-5" role="status">
    <span className="sr-only">{LOADING}...</span>
  </div>
);

Spinner.defaultProps = defaultSPinnerProps;
export { Spinner };
