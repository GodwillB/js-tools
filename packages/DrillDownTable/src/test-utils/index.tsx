import { ReactWrapper } from 'enzyme';

export const renderTable = (wrap: ReactWrapper, text = '') => {
  wrap
    .find('div.table .tr')
    .forEach((tr, indx) => expect(tr.text()).toMatchSnapshot(`${text} tr index ${indx}`));
};
