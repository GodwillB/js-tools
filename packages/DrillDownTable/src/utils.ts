export function columnsFromObject(item: object) {
    // Return columns from item object keys
    const columnsList = [];
    for (const field of Object.keys(item)) {
        const columnItem = {
            Header: field,
            accessor: field
        };
        columnsList.push(columnItem);
    }
    return columnsList;
}
