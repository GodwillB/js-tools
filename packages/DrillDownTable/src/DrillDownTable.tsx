import React, { ComponentType } from 'react';
// Import React Table
import ReactTable, { TableProps } from 'react-table';
import 'react-table/react-table.css';
import { columnsFromObject } from './utils';

function DrillDownTableHOC(WrappedTable: ComponentType) {
  // HOC that enhances ReactTable

  class ResultingTable extends React.Component<Partial<TableProps>, {}> {
    constructor(props: TableProps) {
      super(props);
    }

    public getColumns() {
      // Either use provided columns or get them from the data
      const { data, columns } = this.props;
      if (columns) {
        return columns;
      }
      if (data) {
        return columnsFromObject(data[0]);
      }
    }

    public render() {
      const newProps = {
        columns: this.getColumns()
      };
      return <WrappedTable {...newProps} {...this.props} />;
    }
  }

  return ResultingTable;
}

const DrillDownTable = DrillDownTableHOC(ReactTable);

export default DrillDownTable;
