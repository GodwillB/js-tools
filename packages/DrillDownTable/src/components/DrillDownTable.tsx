import { Dictionary } from '@onaio/utils';
import React, { useState } from 'react';
import { Cell } from 'react-table';
import { DrillDownTableProps } from '../helpers/types';
import { defaultDrillDownFilter } from '../helpers/utils';
import { DropDownCell, DropDownCellProps, Spinner } from './HelperComponents';
import { defaultTableProps, DrillDownColumn, Table, TableJSXProps } from './TableJSX';

/** only provide defaults for the props that are actionable as part of this HOC */
export const defaultDrillDownTableProps = {
  ...defaultTableProps,
  CellComponent: DropDownCell,
  drillDownFilter: defaultDrillDownFilter,
  loading: false,
  loadingComponent: Spinner
};

/** HOC component; wraps around and controls a component that makes use of react-table hooks
 * its main goal is to filter data based on set current parent id  and pass it to controlled presentational component
 */
function DrillDownTable<D extends object>(props: DrillDownTableProps<D>) {
  const {
    columns,
    data,
    parentIdentifierField,
    hasChildren,
    loadingComponent: LoadingComponent
  } = props;
  const parentNodes =
    data && parentIdentifierField ? data.map((el: Dictionary) => el[parentIdentifierField]) : [];
  const [pageData, setPageData] = useState<D[]>([]);

  const mutatedColumns = React.useMemo(() => columns.map(mutateColumns), [columns]) as Array<
    DrillDownColumn<D>
  >;

  /** filters out props.data and excludes records whose parent record does not have the specified id
   * it is invoked each time the presentational component renders only if props.data has changed
   */
  const fetchData = React.useCallback(
    ({ skipPageResetRef, currentParentId: parentId }) => {
      /** a useRef that used to control how react-table resets its state, by default react-table
       * will reset state everytime data option changes, here we want to change the data option
       * that we pass to react-table but we dont want the state to change.
       */
      skipPageResetRef.current = true;
      let filterByLevel = props.data;
      if (props.useDrillDown) {
        filterByLevel = props.drillDownFilter(props, parentId);
      }
      setPageData(filterByLevel);
    },
    [data]
  );

  /** Get modified columns
   * Modify the linker column to include an indicator that you can use to
   * drill-down
   */
  function mutateColumns(el: Dictionary) {
    const { linkerField, CellComponent, extraCellProps } = props;
    if (el.hasOwnProperty('columns') && el.columns && el.columns.length > 0) {
      const newColumns = el.columns.map(mutateColumns);
      el.columns = newColumns;
    }

    if (el.accessor === linkerField) {
      el.Cell = (cell: Cell<D>) => {
        if (CellComponent !== undefined) {
          const { identifierField } = props;

          let thisCellHasChildren = false;
          if (
            hasChildren &&
            identifierField &&
            hasChildren(cell, parentNodes, identifierField, props.data)
          ) {
            thisCellHasChildren = true;
          }

          const cellProps: DropDownCellProps<D> = {
            cell,
            cellValue: cell.value,
            hasChildren: thisCellHasChildren
          };

          if (extraCellProps !== undefined) {
            Object.assign(cellProps, extraCellProps);
          }
          return <CellComponent {...cellProps} />;
        }
        return cell.value;
      };
    }
    return el;
  }

  /** pass all props to the presentational component */
  const TableProps = {
    ...props,
    columns: mutatedColumns,
    data: pageData,
    fetchData,
    parentNodes
  };

  return <>{!props.loading ? <Table {...TableProps} /> : <LoadingComponent />}</>;
}

DrillDownTable.defaultProps = defaultDrillDownTableProps;

export { DrillDownTable };
