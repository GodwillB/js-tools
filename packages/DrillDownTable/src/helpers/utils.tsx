import { DrillDownColumn } from '../components/TableJSX';

/** Takes an object and returns columns that are compatible with react-table
 * derived from the object's keys
 */
export function columnsFromObjects<D extends object>(items: D[]): Array<DrillDownColumn<D>> {
  if (items.length < 1) {
    return [];
  }
  return (Object.keys(items[0]).map(k => ({ Header: k, accessor: k })) as unknown) as Array<
    DrillDownColumn<D>
  >;
}
