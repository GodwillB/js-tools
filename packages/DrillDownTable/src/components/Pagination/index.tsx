/** The default custom pagination component for drillDown v7  */
import { Dictionary } from '@onaio/utils';
import React from 'react';
import {
  NEXT,
  OF,
  PAGE,
  PAGE_SIZE_CATEGORIES,
  PREVIOUS,
  ROWS_TO_DISPLAY
} from '../../helpers/constants';
import { ActualTableInstanceProps } from '../TableJSX';
import './pagination.css';

/** interface describes props for Reveal Custom Pagination */
export interface PaginationProps<T extends object> extends ActualTableInstanceProps<T> {
  pageSizeCategories: number[] /** an array of page size options */;
  nextText: string;
  rowsToDisplayText: string;
  pageText: string;
  ofText: string;
  previousText: string;
}

/** default props for Reveal Pagination */
const defaultPaginationProps = {
  nextText: NEXT,
  ofText: OF,
  pageSizeCategories: PAGE_SIZE_CATEGORIES,
  pageText: PAGE,
  previousText: PREVIOUS,
  rowsToDisplayText: ROWS_TO_DISPLAY
};

/** Reveal pagination component */
function RevealPagination<T extends object = Dictionary>(props: PaginationProps<T>) {
  const {
    gotoPage,
    canPreviousPage,
    canNextPage,
    nextPage,
    previousPage,
    pageOptions,
    setPageSize,
    state: { pageSize, pageIndex },
    pageSizeCategories
  } = props;

  const onChangePageSize = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setPageSize(Number(e.target.value));
  };
  const onClickPrevious = () => {
    previousPage();
  };
  const onChangePageIndex = (e: React.ChangeEvent<HTMLInputElement>) => {
    const page = e.target.value ? Number(e.target.value) - 1 : 0;
    gotoPage(page);
  };
  const onClickNext = () => {
    nextPage();
  };

  return (
    <div className="pagination">
      <span className="mr-2">{props.rowsToDisplayText}</span>
      <select className="page-sizes-select mr-4" value={pageSize} onChange={onChangePageSize}>
        {pageSizeCategories.map(pgSize => (
          <option key={pgSize} value={pgSize}>
            {pgSize}
          </option>
        ))}
      </select>
      <button className="mr-2" onClick={onClickPrevious} disabled={!canPreviousPage}>
        {props.previousText}
      </button>
      <span>
        {props.pageText} {'  '}
        <input
          type="text"
          value={pageIndex + 1}
          onChange={onChangePageIndex}
          style={{ width: '40px' }}
        />{' '}
        {props.ofText} {pageOptions.length}
      </span>
      <button className="ml-2" onClick={onClickNext} disabled={!canNextPage}>
        {props.nextText}
      </button>{' '}
    </div>
  );
}

RevealPagination.defaultProps = defaultPaginationProps;

export { RevealPagination };

/** default function that can be used as a render prop. to use a custom pagination
 * component, create a custom renderProp like this and return your custom pagination from within
 * @param {ActualTableInstanceProps<t>} props - the React table instance containing applied hooks.
 */
export const renderPaginationFun = <T extends object>(props: ActualTableInstanceProps<T>) => {
  return <RevealPagination {...props} />;
};
