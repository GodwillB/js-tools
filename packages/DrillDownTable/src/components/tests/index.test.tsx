import { Dictionary } from '@onaio/utils/dist/types/types';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Column } from 'react-table';
import { columnsFromObjects } from '../../helpers/utils';
import { renderTable } from '../../test-utils';
import { DrillDownTable } from '../DrillDownTable';
import { DropDownCellProps } from '../HelperComponents';
import { renderPaginationFun } from '../Pagination';
import { RenderFiltersInBarOptions } from '../TableJSX';
import { data, dataLowestLevel, jurisdictions } from './fixtures';

const dataColumns = columnsFromObjects(data);
const jurisdictionColumns = columnsFromObjects(jurisdictions);
const emptyColumns = columnsFromObjects([]);

describe('DrillDownTable', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const props = {
      columns: emptyColumns,
      data: [],
      linkerField: 'location'
    };
    shallow(<DrillDownTable {...props} />);
  });

  it('renders correctly even with invalid linkerField', () => {
    const props = {
      columns: dataColumns,
      data,
      linkerField: 'fakeColumn',
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);

    expect(wrapper.find('Table').props().data).toEqual(data.filter(e => e.parent_id === null));
    wrapper.unmount();
  });

  it('renders correctly with derived columns', () => {
    const props = {
      columns: dataColumns,
      data,
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    expect((wrapper.find('Table').props() as any).columns).toMatchSnapshot([
      { Cell: expect.any(Function), Header: 'id', accessor: 'id' },
      { Header: 'location', accessor: 'location' },
      {
        Header: 'parent_id',
        accessor: 'parent_id'
      },
      { Header: 'spray_coverage', accessor: 'spray_coverage' },
      { Header: 'spray_effectiveness', accessor: 'spray_effectiveness' }
    ]);
    wrapper.unmount();
  });

  it('renders correctly with custom columns', () => {
    const columns = [
      {
        Header: 'Name',
        accessor: 'location'
      },
      {
        Header: 'ID',
        accessor: 'id'
      },
      {
        Header: 'Parent ID',
        accessor: 'parent_id'
      },
      {
        Header: 'Spray Coverage',
        accessor: 'spray_coverage'
      }
    ];
    const props: any = {
      columns,
      data
    };
    // this does not have a useDrilDown it will render all the rows, subject to pagination
    const wrapper = mount(<DrillDownTable {...props} />);
    expect((wrapper.find('Table').props() as any).columns).toMatchSnapshot(columns);
    wrapper.unmount();
  });

  it('click to drill down works', () => {
    const props = {
      columns: dataColumns,
      data,
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    // wrapper.update();
    // render the whole table

    renderTable(wrapper, 'InitialRender');

    // drill down first level
    expect(toJson(wrapper.find('.dd-linker-item.dd-clickable'))).toMatchSnapshot();
    expect(wrapper.find('.dd-linker-item.dd-clickable').length).toEqual(3);
    wrapper
      .find('.dd-linker-item.dd-clickable')
      .first()
      .simulate('click');
    wrapper.update();
    renderTable(wrapper, 'After first drillDown');
    // drill down second level
    expect(wrapper.find('.dd-linker-item.dd-clickable').length).toEqual(2);
    wrapper
      .find('.dd-linker-item.dd-clickable')
      .first()
      .simulate('click');
    wrapper.update();
    // render the whole table
    renderTable(wrapper, 'After second Drilldown');
    // there should now be no more drilling down possible
    expect(wrapper.find('.dd-linker-item.dd-clickable').length).toEqual(0);
    wrapper.unmount();
  });

  it('renders correctly lowest level hierarchy', () => {
    const props = {
      columns: dataColumns,
      data: dataLowestLevel,
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.text()).toMatchSnapshot('full rendered data');
    wrapper.unmount();
  });

  it('renders correctly with custom columns and custom linker column', () => {
    const columns: Array<Column<Dictionary>> = [
      {
        Header: 'Name',
        accessor: 'location'
      },
      {
        Header: 'ID',
        accessor: 'id'
      },
      {
        Header: 'Parent ID',
        accessor: 'parent_id'
      },
      {
        Header: 'Spray Coverage',
        accessor: 'spray_coverage'
      }
    ];
    const props = {
      columns,
      data,
      linkerField: 'location',
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    renderTable(wrapper, 'Full render');
    wrapper.unmount();
  });

  it('works fine with useDrillDown being flase', () => {
    const props = {
      columns: dataColumns,
      data,
      useDrillDown: false
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.text()).toMatchSnapshot('Full render');
    wrapper.unmount();
  });

  it('works fine with extraCellProps', () => {
    /** Interface for cell props */
    interface NewCellProps extends DropDownCellProps {
      urlPath: string;
      caret: string;
    }

    /** Custom cell component for testing.
     */
    const NewCell: React.ElementType = (props: NewCellProps) => {
      const { cellValue, hasChildren, urlPath, caret } = props;
      return (
        <div>
          <span>
            {hasChildren ? (
              <a href={urlPath}>
                {cellValue} {caret}
              </a>
            ) : (
              cellValue
            )}
          </span>
        </div>
      );
    };

    const cellProps = {
      CellComponent: NewCell,
      columns: dataColumns,
      data,
      extraCellProps: { urlPath: 'http://example.com', caret: <span>&#43;</span> }
    };
    const wrapper = mount(<DrillDownTable {...cellProps} />);
    expect(wrapper.text()).toMatchSnapshot('find the carret');
    wrapper.unmount();
  });
  it('renders correctly without data', () => {
    const columns = [
      {
        Header: 'Name',
        accessor: 'location'
      },
      {
        Header: 'ID',
        accessor: 'id'
      },
      {
        Header: 'Parent ID',
        accessor: 'parent_id'
      },
      {
        Header: 'Spray Coverage',
        accessor: 'spray_coverage'
      }
    ];
    const props: any = {
      columns,
      data: []
    };
    // should render the default null data component
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.text()).toMatchInlineSnapshot(`"NameIDParent IDSpray CoverageNo Data Found"`);
    wrapper.unmount();
  });
  it('renders custom null data component correctly', () => {
    const textNode =
      'You start forgetting the things you should remember and remembering the things you should forget.';
    const CustomNullData = () => <div id="#ghost">{textNode}</div>;
    const columns = [
      {
        Header: 'Name',
        accessor: 'location'
      },
      {
        Header: 'ID',
        accessor: 'id'
      },
      {
        Header: 'Parent ID',
        accessor: 'parent_id'
      },
      {
        Header: 'Spray Coverage',
        accessor: 'spray_coverage'
      }
    ];
    const props: any = {
      columns,
      data: [],
      nullDataComponent: CustomNullData
    };
    // should render the default null data component
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"NameIDParent IDSpray CoverageYou start forgetting the things you should remember and remembering the things you should forget."`
    );
    wrapper.unmount();
  });
  it('respects loading prop', () => {
    const props = {
      columns: dataColumns,
      data,
      loading: true,
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.find('Spinner').length).toEqual(1);
    wrapper.unmount();
  });

  it('renders custom loading correctly', () => {
    const textNode = 'It is fatal to enter any war without the will to win it.';
    const CustomLoading = () => <div id="#MacArthur">{textNode}</div>;
    const props = {
      columns: dataColumns,
      data,
      loading: true,
      loadingComponent: CustomLoading,
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"It is fatal to enter any war without the will to win it."`
    );
    wrapper.unmount();
  });
  // pagination stuff.
  it('pagination works correctly', () => {
    /** default renderInFilter Function; includes the pagination, customize columns, and row height */
    const customRenderInFilterBar = <T extends object>(
      tableProps: RenderFiltersInBarOptions<T>
    ) => {
      return (
        <div className="row">
          <div className="col">{renderPaginationFun(tableProps)}</div>
        </div>
      );
    };
    const props = {
      columns: jurisdictionColumns,
      data: jurisdictions,
      linkerField: 'name',
      renderInTopFilterBar: customRenderInFilterBar,
      rootParentId: '',
      useDrillDown: true
    };
    const wrapper = mount(<DrillDownTable {...props} />);
    /** pagination components are rendered */
    expect(wrapper.find('RevealPagination').text()).toMatchInlineSnapshot(
      `"Rows to display10203050PreviousPage    Of 2Next "`
    );

    wrapper.update();
    /** contents rendered on page 1; should be about 10 */
    renderTable(wrapper, 'page 1 ');

    // got to next page.
    const buttonsNum = wrapper.find('button').length;
    wrapper
      .find('button')
      .at(buttonsNum - 1)
      .simulate('click');
    wrapper.update();
    renderTable(wrapper, 'page 2 ');
    wrapper.unmount();
  });
});
