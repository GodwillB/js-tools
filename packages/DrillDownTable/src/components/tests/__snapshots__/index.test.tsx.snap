// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrillDownTable click to drill down works 1`] = `
Array [
  <div
    className="dd-linker-item dd-clickable"
  >
    <span>
      1
      <span
        className="dd-caret"
      >
         ▼
      </span>
    </span>
  </div>,
  <div
    className="dd-linker-item dd-clickable"
  >
    <span>
      2
      <span
        className="dd-caret"
      >
         ▼
      </span>
    </span>
  </div>,
  <div
    className="dd-linker-item dd-clickable"
  >
    <span>
      3
      <span
        className="dd-caret"
      >
         ▼
      </span>
    </span>
  </div>,
]
`;

exports[`DrillDownTable click to drill down works: After first drillDown tr index 0 1`] = `"idlocationparent_idspray_coveragespray_effectiveness"`;

exports[`DrillDownTable click to drill down works: After first drillDown tr index 1 1`] = `"4 ▼HFC 1180%80%"`;

exports[`DrillDownTable click to drill down works: After first drillDown tr index 2 1`] = `"5 ▼HFC 2180%80%"`;

exports[`DrillDownTable click to drill down works: After second Drilldown tr index 0 1`] = `"idlocationparent_idspray_coveragespray_effectiveness"`;

exports[`DrillDownTable click to drill down works: After second Drilldown tr index 1 1`] = `"9Operational Area 9470%90%"`;

exports[`DrillDownTable click to drill down works: After second Drilldown tr index 2 1`] = `"10Operational Area 10480%100%"`;

exports[`DrillDownTable click to drill down works: After second Drilldown tr index 3 1`] = `"11Operational Area 114100%100%"`;

exports[`DrillDownTable click to drill down works: InitialRender tr index 0 1`] = `"idlocationparent_idspray_coveragespray_effectiveness"`;

exports[`DrillDownTable click to drill down works: InitialRender tr index 1 1`] = `"1 ▼District A80%80%"`;

exports[`DrillDownTable click to drill down works: InitialRender tr index 2 1`] = `"2 ▼District B75%85%"`;

exports[`DrillDownTable click to drill down works: InitialRender tr index 3 1`] = `"3 ▼District C90%90%"`;

exports[`DrillDownTable click to drill down works: InitialRender tr index 4 1`] = `"18District D0%0%"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 0 1`] = `"idgeographic_levelparent_idname"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 1 1`] = `"64301afa-e973-447b-a88c-4da20025c76f1Trat ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 2 1`] = `"7f204867-fab0-4246-a97c-92e0b936cab61Ubon Ratchathani ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 3 1`] = `"61707fc2-c6ac-4112-a8d6-2a48619583961Trat 2 ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 4 1`] = `"29400Katete ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 5 1`] = `"29420Lusaka ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 6 1`] = `"f8863022-ff88-4c22-b2d1-83f59f31b8740Oddar Meanchey Province ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 7 1`] = `"29390Chadiza ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 8 1`] = `"9c3c2db4-bddd-44c5-870a-a0eef539e4da0Lop Buri ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 9 1`] = `"29410Sinda ▼"`;

exports[`DrillDownTable pagination works correctly: page 1  tr index 10 1`] = `"39530Siavonga"`;

exports[`DrillDownTable pagination works correctly: page 2  tr index 0 1`] = `"idgeographic_levelparent_idname"`;

exports[`DrillDownTable pagination works correctly: page 2  tr index 1 1`] = `"39540Siavonga ▼"`;

exports[`DrillDownTable renders correctly lowest level hierarchy: full rendered data 1`] = `"idlocationparent_idspray_coveragespray_effectivenessNo Data Found"`;

exports[`DrillDownTable renders correctly with custom columns 1`] = `
Object {
  "0": Object {
    "Header": "Name",
    "accessor": "location",
  },
  "1": Object {
    "Cell": [Function],
    "Header": "ID",
    "accessor": "id",
  },
  "2": Object {
    "Header": "Parent ID",
    "accessor": "parent_id",
  },
  "3": Object {
    "Header": "Spray Coverage",
    "accessor": "spray_coverage",
  },
}
`;

exports[`DrillDownTable renders correctly with custom columns and custom linker column: Full render tr index 0 1`] = `"NameIDParent IDSpray Coverage"`;

exports[`DrillDownTable renders correctly with custom columns and custom linker column: Full render tr index 1 1`] = `"District A ▼180%"`;

exports[`DrillDownTable renders correctly with custom columns and custom linker column: Full render tr index 2 1`] = `"District B ▼275%"`;

exports[`DrillDownTable renders correctly with custom columns and custom linker column: Full render tr index 3 1`] = `"District C ▼390%"`;

exports[`DrillDownTable renders correctly with custom columns and custom linker column: Full render tr index 4 1`] = `"District D180%"`;

exports[`DrillDownTable renders correctly with derived columns 1`] = `
Object {
  "0": Object {
    "Cell": [Function],
    "Header": "id",
    "accessor": "id",
  },
  "1": Object {
    "Header": "location",
    "accessor": "location",
  },
  "2": Object {
    "Header": "parent_id",
    "accessor": "parent_id",
  },
  "3": Object {
    "Header": "spray_coverage",
    "accessor": "spray_coverage",
  },
  "4": Object {
    "Header": "spray_effectiveness",
    "accessor": "spray_effectiveness",
  },
}
`;

exports[`DrillDownTable works fine with extraCellProps: find the carret 1`] = `"idlocationparent_idspray_coveragespray_effectiveness1 +District A80%80%2 +District B75%85%3 +District C90%90%18District D0%0%"`;

exports[`DrillDownTable works fine with useDrillDown being flase: Full render 1`] = `"idlocationparent_idspray_coveragespray_effectiveness1 ▼District A80%80%4 ▼HFC 1180%80%9Operational Area 9470%90%10Operational Area 10480%100%11Operational Area 114100%100%5 ▼HFC 2180%80%12Operational Area 12586%100%2 ▼District B75%85%6 ▼HFC 3280%80%17HFC 1720%0%"`;
